src/kernel/tracelog.c

```diff
@@ -0,0 +1,568 @@
+            } else if (buffer_compare_with_cstring(relative_uri, "clear")) {
+                tracelog_collate();
+                tracelog_clear();
+                tracelog_send_http_simple_result(handler, "tracelog cleared");
+            } else {
+                tracelog_send_http_uri_not_found(handler);
+            }
+            break;
+        }
+        tracelog_collate();
+        tracelog_send_http_chunked_response(handler);
+        if (tracelog_do_http_get(handler, relative_uri)) {
+            init_closure(&tracelog.send_http_chunk, tracelog_send_http_chunk, handler, relative_uri);
+            schedule_send_http_chunk();
+        }
+        break;
+    case HTTP_REQUEST_METHOD_PUT:
+        /* not suported */
+    default:
+        tracelog_send_http_no_method(handler);
+        break;
+    }
+}
+
+static void init_tracelog_http_listener(void)
+{
+    tracelog.http_listener = allocate_http_listener(tracelog.h, TRACELOG_HTTP_PORT);
+    assert(tracelog.http_listener != INVALID_ADDRESS);
+    http_register_uri_handler(tracelog.http_listener,
+                              TRACELOG_TRACE_URI,
+                              closure(tracelog.h, tracelog_http_request));
+
+    connection_handler ch = connection_handler_from_http_listener(tracelog.http_listener);
+    status s = listen_port(tracelog.h, TRACELOG_HTTP_PORT, ch);
+    if (is_ok(s)) {
+        rprintf("started tracelog http listener on port %d\n", TRACELOG_HTTP_PORT);
+    } else {
+        msg_err("listen_port() (port %d) failed for tracelog HTTP listener\n",
+                TRACELOG_HTTP_PORT);
+        deallocate_http_listener(tracelog.h, tracelog.http_listener);
+    }
+}
+
+void init_tracelog_config(tuple root)
+{
+    tuple tl = get(root, sym(tracelog));
+    if (!tl)
+        return;
+
+    if (get(tl, sym(disable))) {
+        /* don't trace on startup */
+        tracelog.disabled = true;
+        rprintf("tracelog disabled on start\n");
+    }
+
+    bytes alloc_size;
+    if (get_u64(tl, sym(alloc_size), &alloc_size))
+        tracelog.alloc_size = alloc_size;
+
+    tracelog.trace_tags = get_tuple(tl, sym(trace_tags));
+}
+
+void init_tracelog(heap h)
+{
+    tracelog.h = h;
+    tracelog.m = allocate_mutex(h, 0 /* no spinning */);
+    assert(tracelog.m != INVALID_ADDRESS);
+    list_init(&tracelog.collated_entries);
+    list_init(&tracelog.free_tracelog_buffers);
+    tracelog.n_free = 0;
+    tracelog.alloc_size = TRACELOG_DEFAULT_BUFFER_SIZE;
+    tracelog.trace_tags = 0;
+    init_closure(&tracelog.collator, tracelog_collator);
+    tracelog.collator_scheduled = false;
+    tracelog.disabled = false;
+    
+    for (int i = 0; i < present_processors; i++) {
+        tracelog_buffer b = allocate_tracelog_buffer_locked(); /* don't need lock here */
+        assert(b != INVALID_ADDRESS);
+        cpuinfo ci = cpuinfo_from_id(i);
+        ci->tracelog_buffer = b;
+        vector cv = allocate_vector(h, 8);
```

Comment:
What's the purpose of this vector allocation?

Reply:
This was left-over cruft from an earlier implementation. I removed it.